File: JasSwift/Sources/Panels/StrokePanel.swift
/// Stroke panel menu definition.

public enum StrokePanel {
    /// Source of truth is workspace/panels/stroke.yaml's `menu:` block
    /// (review #15); the generic reader builds the items from the bundle.
    /// The cap/join radio rows share one YAML action each
    /// (`set_stroke_cap` / `set_stroke_join`), so the builder folds each
    /// `params.cap` / `params.join` into the command
    /// (`set_stroke_cap:butt`, …); `dispatch` splits that suffix back off.
    public static func menuItems() -> [PanelMenuItem] {
        menuItemsFromYaml("stroke_panel_content")
    }

    public static func dispatch(_ cmd: String, addr: PanelAddr,
                                layout: inout WorkspaceLayout,
                                model: Model? = nil) {
        if cmd == "close_panel" { layoutApply(&layout, opClosePanel(addr)); return }
        // Cap/join radios arrive param-folded from the generic menu
        // builder (`set_stroke_cap:round`); split the value back off and
        // route through the YAML action so both panel state (panel.cap)
        // and global state (stroke_cap, which propagates to the
        // selection) update — matching the in-panel cap/join buttons.
        guard let model = model else { return }
        if let cap = strip(cmd, prefix: "set_stroke_cap:") {
            runYamlActionByName("set_stroke_cap", params: ["cap": cap], model: model)
            return
        }
        if let join = strip(cmd, prefix: "set_stroke_join:") {
            runYamlActionByName("set_stroke_join", params: ["join": join], model: model)
            return
        }
    }

    private static func strip(_ s: String, prefix: String) -> String? {
        s.hasPrefix(prefix) ? String(s.dropFirst(prefix.count)) : nil
    }
}

File: jas_dioxus/src/panels/stroke_panel.rs
//! Stroke panel menu definition.

use crate::workspace::app_state::AppState;
use crate::workspace::workspace::PanelAddr;
use super::panel_menu::PanelMenuItem;

/// Menu items for the Stroke panel.
pub fn menu_items() -> Vec<PanelMenuItem> {
    // Source of truth is workspace/panels/stroke.yaml's `menu:` block
    // (review #15); the generic reader builds the items from the bundle.
    super::panel_menu::menu_items_from_yaml("stroke_panel_content")
}

/// Dispatch a menu command for the Stroke panel.
///
/// The cap/join radio commands arrive param-folded from the generic
/// menu builder (`set_stroke_cap:round`, `set_stroke_join:bevel`) — see
/// `panel_menu::command_with_params`. We split the suffix back off and
/// write the panel-state field, then push to the selection so the menu
/// and the in-panel cap/join buttons stay in sync.
pub fn dispatch(cmd: &str, addr: PanelAddr, state: &mut AppState) {
    if let Some(cap) = cmd.strip_prefix("set_stroke_cap:") {
        state.stroke_panel.cap = cap.to_string();
        state.apply_stroke_panel_to_selection();
        return;
    }
    if let Some(join) = cmd.strip_prefix("set_stroke_join:") {
        state.stroke_panel.join = join.to_string();
        state.apply_stroke_panel_to_selection();
        return;
    }
    match cmd {
        "close_panel" => state.workspace_layout.close_panel(addr),
        _ => {}
    }
}

/// Query whether a radio command is checked. The cap/join radio
/// commands carry their value as a `:suffix` (see `dispatch`); the
/// checkmark follows the matching panel-state field.
pub fn is_checked(cmd: &str, state: &AppState) -> bool {
    if let Some(cap) = cmd.strip_prefix("set_stroke_cap:") {
        return state.stroke_panel.cap == cap;
    }
    if let Some(join) = cmd.strip_prefix("set_stroke_join:") {
        return state.stroke_panel.join == join;
    }
    false
}
